### ISSUE TITLE
PyTorch usage of INMSLayer on TensorRT

### ISSUE THREAD

**Issue by levipereira** [NVIDIA/TensorRT#3807]
I need help implementing the INMSLayer and stop using the Efficient NMS Plugin. Do you have any implementation examples for this layer INMSLayer  in Python and PyTorch/Onnx?

**Comment by lix19937** [comment]
you can use torchvision.ops.nms define a nn.module, then export a onnx or you can build your own network by api    
```
opt_profile = get_runner.builder.create_optimization_profile()
get_runner.config.add_optimization_profile(opt_profile)

boxes = network.add_input("boxes", dtype=trt.float32, shape=(1, 3, 4))
scores = network.add_input("scores", dtype=trt.float32, shape=(1, 3, 3))
constant = network.add_constant(shape=(), weights=np.ones(shape=(), dtype=np.int32))
max_output_boxes_per_class = constant.get_output(0)

layer = network.add_nms(boxes, scores, max_output_boxes_per_class)
layer.get_output(0).dtype = trt.int32
layer.get_output(1).dtype = trt.int32
network.mark_output(layer.get_output(0))
network.mark_output(layer.get_output(1))

inputs[boxes.name] = np.array([[[0.0, 0.0, 0.1, 0.1], [0.2, 0.2, 0.4, 0.4], [0.5, 0.5, 0.6, 0.6]]])
inputs[scores.name] = np.array([[[0.7, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.9]]])

# Expected shape is [2, 3]
outputs[layer.get_output(0).name] = layer.get_output(0).shape
expected[layer.get_output(0).name] = np.array([[0, 2, 2], [0, 0, 0]])

# Expected shape is [] with a scalar value of 2
outputs[layer.get_output(1).name] = layer.get_output(1).shape
expected[layer.get_output(1).name] = np.array(2)
```  
also INMSLayer has some limits    

> There is a hardware-dependent limit K such that only the K highest scoring boxes in each batch item will be considered for selection. The value of K is 2000 for SM 5.3 and 6.2 devices, and 5000 otherwise.
> Do not inherit from this class, as doing so will break forward-compatibility of the API and ABI.  

https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_n_m_s_layer.html    
https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Graph/Layers.html#inmslayer

**Comment by levipereira** [comment]
@lix19937 Thank for you example. 
I must be missing something.
I didn't understand how to implement two different types of Non-Maximum Suppression (NMS) in the ONNX graph, using `torchvision.ops.nms` and `network.add_nms`.
I need to implement the INMSLayer in this [part of the code](https://github.com/WongKinYiu/yolov9/blob/93f1a2829870a37b306115012f34804017cba01b/models/experimental.py#L213). 
If you could provide a simple example, it would be really helpful.

**Comment by levipereira** [comment]
I understand the TensorRT API implementation. 
The example I need is how to implement NMS (INMSLayer)  in the ONNX Graph. I want to use trtexec to generate the engine with the INMSLayer layer instead of EfficientNMS.

Is this possible? The ultimate goal is to generate the Engine with trtexec using a Pytorch-to-ONNX exported model with the INMSLayer layer.

The challenge is how to implement INMS  on graph ONNX only. Later will use trtexec to build engine.

**Comment by lix19937** [comment]
> Is this possible? The ultimate goal is to generate the Engine with trtexec using a Pytorch-to-ONNX exported model with the INMSLayer layer.   

If you split the model into two parts, and the nms op use trt-api,and connect two parts, otherwise no possible.

**Comment by levipereira** [comment]
I found the answer was use [onnx-tensorrt](https://github.com/onnx/onnx-tensorrt).
 [the importer definition for NonMaxSuppression](https://github.com/onnx/onnx-tensorrt/blob/6ba67d3428e05f690145373ca87fb8d32f98df45/builtin_op_importers.cpp#L3122-L3205) in ONNX parser for an example of INMSLayer usage.

**Comment by korkland** [comment]
@lix19937 Hi, ive tried to use torchvision.ops.nms with the following implementation and got Cast error, any idea how can i make it work?

`    def forward(self, boxes, scores, iou_threshold: float, max_count: int):
        keep = torchvision.ops.nms(boxes, scores, iou_threshold)[:max_count]
        pad_count = max_count - keep.size(0)
        return torch.nn.functional.pad(keep, (0, pad_count), value=keep[-1])`

using TensorRT-10.1.0.27

`[07/21/2024-07:24:56] [E] Error[9]: Error Code: 9: Skipping tactic 0x0000000000000000 due to exception Assertion engine failed. 
[07/21/2024-07:24:56] [E] Error[10]: Error Code: 10: Could not find any implementation for node {ForeignNode[/interpret_2d/Constant_589_output_0.../interpret_2d/nms/nms/Cast]}.
[07/21/2024-07:24:56] [E] Error[10]: IBuilder::buildSerializedNetwork: Error Code 10: Internal Error (Could not find any implementation for node {ForeignNode[/interpret_2d/Constant_589_output_0.../interpret_2d/nms/nms/Cast]}.)`

**Comment by lix19937** [comment]
@korkland  Can you upload the onnx ?

**Comment by korkland** [comment]
@lix19937 Thanks for the quick response, the model is propriety so i cant, but basically when im trying to build the engine with --fp16 it succeeded, when im trying to build it with --int8 (im explicitly quantizing the backbone using modelopt lib, so the rest is fp32) it succeeded, but when i try to build it with --int8 --fp16 im getting the error ive mentioned.

```
data_loader = DataLoader(dataset, batch_size=1, shuffle=False, collate_fn=lambda x: x[0][0])

    # Quantize the model and perform calibration (PTQ)
    # CNN networks only supports INT8_DEFAULT_CFG
    config = copy.deepcopy(mtq.INT8_DEFAULT_CFG)
    config["quant_cfg"]["backbone.*"] = {"enable": True}
    config["quant_cfg"]["head_2d.*"] = {"enable": False}
    config["quant_cfg"]["*interpret_2d.*"] = {"enable": False}
    config["quant_cfg"]["head_3d.*"] = {"enable": False}
    quantized_model = mtq.quantize(model, config, lambda model: [model(x) for x in tqdm(data_loader)])

    # Print quantization summary after successfully quantizing the model with mtq.quantize
    # This will show the quantizers inserted in the model and their configurations
    mtq.print_quant_summary(quantized_model)

    # Export to ONNX
    input_keys, output_keys, const_folding, opset_vers = get_onnx_export_args(quantized_model, inputs_converted,
                                                                              network_name, module_name)
    input = inputs_converted[0]
    opset_version = 17
    empty_kwargs = dict()
    args_for_onnx = tuple([input, empty_kwargs])

    torch.onnx.export(quantized_model, args_for_onnx, path,
                      do_constant_folding=const_folding,
                      opset_version=opset_version,
                      input_names=input_keys,
                      output_names=output_keys,
                      verbose=args.nsp_profile)
```
should it works with torchvision nms? i dont want it to try to quantize the nms

**Comment by laugh12321** [comment]
@lix19937 I generated two YOLOv8s ONNX models using INMSLayer and EfficientNMS respectively, ensuring that their input and output remained consistent.

**INMSLayer**
![image](https://github.com/user-attachments/assets/6205c370-c7e1-4a46-8947-9ce202c8d49b)

**EfficientNMS**
![image](https://github.com/user-attachments/assets/29617529-509b-41e1-bcd5-48ff54ca4f9c)

Then, I converted these models to engines using `trtexec` and tested the inference time. The results showed that while their accuracy was almost identical, the engine exported from INMSLayer was significantly slower than the one exported from EfficientNMS. This could be due to additional processing needed to keep the output nodes consistent. Additionally, EfficientNMS has a `class_agnostic` attribute that INMSLayer lacks, which can be very useful in certain applications.

Overall, EfficientNMS has a clear advantage over INMSLayer, except for its inability to return indices. 

May I ask why there is consideration to deprecate EfficientNMS?

**Comment by lix19937** [comment]
EfficientNMSONNXPlugin is deprecated since TensorRT 9.0. Its functionality has been superseded by the INMSLayer.   

The NMS algorithm iterates through a set of bounding boxes and their confidence scores, in decreasing order of score. Boxes are selected if their score is above a given threshold, and their intersection-over-union (IoU) with previously selected boxes is less than or equal to a given threshold. This layer implements NMS per batch item and per class.

Per batch item, boxes are initially sorted by their scores without regard to class. Only boxes up to a maximum of the TopK limit are considered for selection (per batch). During selection, only overlapping boxes of the same class are compared, so that overlapping boxes of different classes do not suppress each other.

**Comment by laugh12321** [comment]
@lix19937 Thanks for the clarification. So, if I want overlapping bounding boxes of different classes to suppress each other, INMSLayer cannot achieve this? Only EfficientNMSONNXPlugin with `class_agnostic` set to `true` can do this?

**Comment by lix19937** [comment]
INMSLayer support per class.

**Comment by laugh12321** [comment]
@lix19937 Thanks for the clarification.

**Comment by korkland** [comment]
> Per batch item, boxes are initially sorted by their scores without regard to class. Only boxes up to a maximum of the TopK limit are considered for selection (per batch). During selection, only overlapping boxes of the same class are compared, so that overlapping boxes of different classes do not suppress each other.
> 
> ref https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_n_m_s_layer.html

can you please share your code of both NMS insertions?

Thanks

**Comment by laugh12321** [comment]
@korkland You can find the code for inserting EfficientNMS at the following link: https://github.com/laugh12321/TensorRT-YOLO/blob/v4.0/tensorrt_yolo/export/head.py#L159C1-L190C10

For INMSLayer insertion, I made some modifications based on the following link to ensure the output is consistent with EfficientNMS: https://github.com/meituan/YOLOv6/blob/e86a483f3f6bded25d45970b56831345a99744a4/yolov6/models/end2end.py#L140C1-L189C59.

**Comment by duong0411** [comment]
i have a question , i converted file pt yolov8 to onnx end2end model with Efficient NMS. And i convert model onnx to tensorrt with trtexec tool and fp16. Why my accuuracy is negative ?

**Comment by duong0411** [comment]
![Image](https://github.com/user-attachments/assets/6445c224-6b4e-4627-ada9-53fdc08aa82b)
